src/main/core/support/configuration.rs

```diff
@@ -1,9 +1,517 @@
+static GENERAL_HELP: Lazy<std::collections::HashMap<String, String>> =
+    Lazy::new(|| generate_help_strs(schema_for!(GeneralOptions)));
+
+// these must all be Option types since they aren't required by the CLI, even if they're
+// required in the configuration file
+#[derive(Debug, Clone, Clap, Serialize, Deserialize, Merge, JsonSchema)]
+#[clap(help_heading = "GENERAL (Override configuration file options)")]
+#[serde(deny_unknown_fields)]
+pub struct GeneralOptions {
+    /// The simulated time at which simulated processes are sent a SIGKILL signal
+    #[clap(long, value_name = "seconds")]
+    #[clap(about = GENERAL_HELP.get("stop_time").unwrap())]
+    stop_time: Option<units::Time<units::TimePrefixUpper>>,
+
+    /// Environment variables passed to all simulated processes (ex: "ENV_A=1;ENV_B=2")
+    #[clap(long, value_name = "env-string")]
+    #[clap(about = GENERAL_HELP.get("environment").unwrap())]
+    environment: Option<String>,
+
+    /// Initialize randomness for each thread using seed N
+    #[clap(long, value_name = "N")]
+    #[clap(about = GENERAL_HELP.get("seed").unwrap())]
+    #[serde(default = "default_some_1")]
+    seed: Option<u32>,
+
+    /// Run concurrently with N worker threads
+    #[clap(long, short = 'w', value_name = "N")]
+    #[clap(about = GENERAL_HELP.get("workers").unwrap())]
+    #[serde(default = "default_some_0")]
+    workers: Option<u32>,
+
+    #[clap(long, value_name = "seconds")]
+    #[clap(about = GENERAL_HELP.get("bootstrap_end_time").unwrap())]
+    #[serde(default = "default_some_time_0")]
+    bootstrap_end_time: Option<units::Time<units::TimePrefixUpper>>,
+
+    /// Log level of output written on stdout. If Shadow was built in release mode, then log
+    /// messages at a level lower than 'info' will always be dropped
+    #[clap(long, short = 'l', value_name = "level")]
+    #[clap(about = GENERAL_HELP.get("log_level").unwrap())]
+    #[serde(default = "default_some_message")]
+    log_level: Option<LogLevel>,
+
+    /// Interval at which to print heartbeat messages
+    #[clap(long, value_name = "seconds")]
+    #[clap(about = GENERAL_HELP.get("heartbeat_interval").unwrap())]
+    #[serde(default = "default_some_time_1")]
+    heartbeat_interval: Option<units::Time<units::TimePrefixUpper>>,
+
+    /// Path to store simulation output
+    #[clap(long, short = 'd', value_name = "path")]
+    #[clap(about = GENERAL_HELP.get("data_directory").unwrap())]
+    #[serde(default = "default_data_directory")]
+    data_directory: Option<String>,
+
+    /// Path to recursively copy during startup and use as the data-directory
+    #[clap(long, short = 'e', value_name = "path")]
+    #[clap(about = GENERAL_HELP.get("template_directory").unwrap())]
+    template_directory: Option<String>,
+}
+
+impl GeneralOptions {
+    /// Replace unset (`None`) values of `base` with values from `default`.
+    pub fn merge(base: &Self, default: &Self) -> Self {
+        let mut base = base.clone();
```

Comment:
Nit - take `base` and `default` by value and let caller `clone` if they still need them? I don't expect a noticeable performance impact here, but it's a useful pattern in general for functions that would otherwise clone their parameters.

Reply:
Added.